MODE-2162: Store indexes in ElasticSearch

okulikov opened the pull request:
This pull request implements Elasticsearch index provider

hchiorean commented on index-providers/modeshape-elasticsearch-index-provider/pom.xml:
```diff
@@ -0,0 +1,49 @@
+
+    <dependencies>
+        <dependency>
+            <groupId>org.elasticsearch</groupId>
```
this should be defined in the parent POM

hchiorean commented on index-providers/modeshape-elasticsearch-index-provider/pom.xml:
```diff
@@ -0,0 +1,49 @@
+        <dependency>
+            <groupId>org.apache.lucene</groupId>
+            <artifactId>lucene-core</artifactId>
+            <version>4.10.4</version>
```
this should be extracted into a property, otherwise it's duplicated across all dependencies

hchiorean commented on index-providers/modeshape-elasticsearch-index-provider/src/main/java/org/modeshape/jcr/index/elasticsearch/EmptyResults.java:
```diff
@@ -0,0 +1,42 @@
+ *
+ * @author kulikov
+ */
+public class EmptyResults implements Index.Results {
```
This class it not needed. Instead, [this](https://github.com/okulikov/modeshape/blob/MODE-2162-b/modeshape-jcr/src/main/java/org/modeshape/jcr/index/local/Operations.java#L57) should be "refactored out" and placed as a constant (`public static final`) in the `Results` interface from the SPI.

hchiorean commented on index-providers/modeshape-elasticsearch-index-provider/src/main/java/org/modeshape/jcr/index/elasticsearch/EsIndexColumns.java:
```diff
@@ -0,0 +1,67 @@
+ */
+public class EsIndexColumns {
+    
+    private Map<String, EsIndexColumn> columns = new HashMap();
```
can this be `final` ?

hchiorean commented on index-providers/modeshape-elasticsearch-index-provider/src/main/java/org/modeshape/jcr/index/elasticsearch/EsIndexProvider.java:
```diff
@@ -0,0 +1,197 @@
+                esNode.stop();
+            }
+        } catch (Exception e) {
+        }
```
shouldn't the exception be logged at least ?

hchiorean commented on index-providers/modeshape-elasticsearch-index-provider/src/main/java/org/modeshape/jcr/index/elasticsearch/EsIndexProvider.java:
```diff
@@ -0,0 +1,197 @@
+            client.close();
+            esNode.stop();
+        } catch (Exception e) {
+        }
```
shouldn't the exception be logged ?

hchiorean commented on index-providers/modeshape-elasticsearch-index-provider/src/main/java/org/modeshape/jcr/index/elasticsearch/EsIndexProvider.java:
```diff
@@ -0,0 +1,197 @@
+ *
+ * @author kulikov
+ */
+public class EsIndexProvider extends IndexProvider {
```
this class should also override the ` protected int getCostEstimate()` method from the base class, making sure it returns a `REMOTE` cost. In other words, if there's a local index (MapDB or Lucene) that covers the same property, it should be preferred over the `REMOTE` one.

hchiorean commented:
In addition to the unit tests specific to the ES provider, there should be an additional test case which extends `LocalIndexProviderTest`. This test case covers a lot of cases which can flush out potential problems with the provider.

This test case _may_ contain additional tests specific to ES: for example testing that when both a local and an ES index is defined over a property, the local index is preferred because of the lower cost. 

The [Lucene index provider](https://github.com/okulikov/modeshape/blob/MODE-2162-b/index-providers/modeshape-lucene-index-provider/src/test/java/org/modeshape/jcr/index/lucene/LuceneIndexProviderTest.java) does a similar thing.

hchiorean commented:
@okulikov: after sorting all the comments (including @rhauch's if he has any), can you please create a separate PR for `4.x` ? 
`master` and `4.x` have already starting diverging significantly and this feature has to be present in both branches. thanks

hchiorean commented:
another question: this PR does not cover any work required for the WF kit integration or the general `modeshape-distribution` integration. Will this be a separte PR ?

okulikov commented:
@hchiorean , `EsIndexProviderTest` extends `LocalIndexProviderTest`

hchiorean commented:
right, sorry I missed it.

rhauch commented:
> another question: this PR does not cover any work required for the WF kit integration or the general `modeshape-distribution` integration. Will this be a separte PR ?

@okulikov, any reply? Also would be great to see at least an acknowledgement on @hchiorean's other line notes, or, if you don't agree with his comment, a line note that explains why.

okulikov commented:
@hchiorean , @rhauch Fixed notes. Let's do WF kit in this PR as well. I just have some issues with testing.

hchiorean commented on index-providers/modeshape-elasticsearch-index-provider/src/main/java/org/modeshape/jcr/index/elasticsearch/EsIndex.java:
```diff
@@ -0,0 +1,327 @@
+    @Override
+    public Results filter(IndexConstraints constraints) {
+        FilterBuilder filter = operations.createFilter(constraints.getConstraints(), constraints.getVariables());
+//        try {
```
shouldn't these comments be removed ?

hchiorean commented on index-providers/modeshape-elasticsearch-index-provider/src/main/java/org/modeshape/jcr/index/elasticsearch/EsIndexProvider.java:
```diff
@@ -0,0 +1,204 @@
+            IndexDefinition defn,
+            NodeTypes.Supplier nodeTypeSupplier,
+            Problems problems) {
+        // first perform some custom validations
```
should this be overridden ? In other words, is it intentional for the ES index provider to not perform any validations ?

okulikov replied:
Lets' keep it as place where to do such validations in the future if it will needed.

hchiorean replied:
ok; I forgot the base class only provides an empty implementation since the default validations are performed always.

okulikov commented:
WF kit and binary distribution part already included into PR

hchiorean commented:
@okulikov: the changes look fine, with some additional comments (see above). 

Also, a couple of more general questions:
1. why isn't the ElasticSearch version the latest one - `2.1.0` (as opposed to `1.7.3`) ? This latest version also uses Lucene 5.3.1, which is exactly what the Lucene index provider uses
2. an additional integration test should be added which validates that the ES provider works in WF. [This](https://github.com/ModeShape/modeshape/blob/master/integration/modeshape-jbossas-integration-tests/src/main/resources/kit/jboss-wf9/standalone/configuration/standalone-modeshape.xml#L330) is where the new index provider should be added and tested. This will most likely require the same test dependencies as the `es-index-provider` module, so they should be added to the `modeshape-jbossas-integration-tests` module and the Arquillian code from the [unit test class](https://github.com/ModeShape/modeshape/blob/master/integration/modeshape-jbossas-integration-tests/src/test/java/org/modeshape/test/integration/QueryIntegrationTest.java#L59) updated.
3. Please rebase the code on the latest `master` or `4.x` because it cannot be automatically merged.

okulikov commented:
@hchiorean answering in the same order:
1. The reason is that older version API provides `Filters` in addition to `Queries` and without using `Filters` I didn't find any way to pass 'all' tests inherited from 'LocalIndexProviderTest`. It very often return ether less results the test case is expecting or more. Let's for now stick to this version and later I can try to reach latest version;
2. The integration test will require access to the ES cluster. If I am right the dependencies should be added to the 'org.elasticsearch' WF module definition. Is it possible to perform it for test purposes only?
3. Will do.

hchiorean commented:
>1. The reason is that older version API provides Filters in addition to Queries and without using Filters I didn't find any way to pass 'all' tests inherited from 'LocalIndexProviderTest`. It very often return ether less results the test case is expecting or more. Let's for now stick to this version and later I can try to reach latest version;

That may be so, but I'm not sure we want to support an older version of Elastic Search, for numerous reasons. Not only there have been a number of bug fixes, but Lucene 5.x is superior to Lucene 4.10
@rhauch thoughts ?

>2. The integration test will require access to the ES cluster. If I am right the dependencies should be added to the 'org.elasticsearch' WF module definition. Is it possible to perform it for test purposes only?

If the test only works with a real cluster, then no, we should definitively not be adding these dependencies to the WF module. I thought that the unit testing done as part of ES index provider artifact uses some sort of in-memory/embedded mode (similar to Cassandra testing for example). If that's not the case, then there's not much more we can do.

okulikov commented:
@hchiorean , @rhauch I can try a bit more to fix problems with latest version. Any thoughts?

hchiorean commented:
my vote is on using the latest Elastic Search version `2.1.0` since this version/API will be evolved in the future by the ES team.

rhauch commented:
@hchiorean, @okulikov: I think we should use the latest version of ES.

> > The integration test will require access to the ES cluster. If I am right the dependencies should be added to the 'org.elasticsearch' WF module definition. Is it possible to perform it for test purposes only?

> If the test only works with a real cluster, then no, we should definitively not be adding these dependencies to the WF module. I thought that the unit testing done as part of ES index provider artifact uses some sort of in-memory/embedded mode (similar to Cassandra testing for example). If that's not the case, then there's not much more we can do.

Actually, if the provider can only be tested with an in-memory ES server, then we should at least do that in this PR. However, if really the only way to test this is with a real ES cluster, we can use Maven to start a Docker container with ES, have the tests use it, and then have Maven shut down the container. Might be better to do that in a separate PR, though. We'll likely want to use Docker in other modules as well -- for example, to test with PostgreSQL, MySQL, or other DBMSes.

okulikov commented:
@hchiorean , @rhauch So I will focus on latest ES version first with PR and then prepare another PR with integration test

rhauch commented:
@okulikov, Sorry if I wasn't clear. I meant that if you can do any integration test with ES without Docker (either because ES can be run in-process or started/managed by the test itself), then that should be done as part of this PR. After all, that's the only way to verify that this feature is working correctly. OTOH, if the only real way to test integration with ES is via Docker, then IMO that can be done in a subsequent PR.

Perhaps you've already seen thins: it appears that the tests can start and manage the ES cluster; see https://www.elastic.co/guide/en/elasticsearch/reference/current/integration-tests.html. If that is the case, that should be the preferred approach, since it will be the simplest (and probably fastest).

okulikov commented:
@rhauch , we are on the same page. I hope that moving to latest version will automatically help to run intergration-test without Docker since all necessary Lucene dependencies  already deployed as WF module.

hchiorean commented:
> OTOH, if the only real way to test integration with ES is via Docker, then IMO that can be done in a subsequent PR.

not sure how easy it would be to integrate Docker with Maven and tie it up together with the build system (while also keeping a full `clean install -Passembly` run in reasonable time). As an alternative, I think we can explore the Docker integration for either the `examples` or `quickstarts` repos

okulikov commented:
One test case remains

okulikov commented:
@hchiorean , @rhauch can we skip one test case `shouldUseIndexesAfterRestarting` ? ES cluster switches to YELLOW state during repository restart.  It was not happening with previous version.

hchiorean commented:
if you mean just for the `ESIndexProviderTest`case, that's fine IMO (i.e. you override this and ignore it just in this particular test case)

okulikov commented:
yes, exactly

hchiorean commented on index-providers/modeshape-elasticsearch-index-provider/pom.xml:
```diff
@@ -0,0 +1,48 @@
+        </dependency>         <!-- 
+            Required to text FTS on binary properties...
+        -->
+        <!-- dependency>
```
if this is not used, it should be removed

okulikov commented:
@rhauch , @hchiorean Done without Docker usage.

hchiorean commented on deploy/jbossas/kit/jboss-wf9/org/elasticsearch/2.1.0/module.xml:
```diff
@@ -0,0 +1,51 @@
+               <include path="sun/misc/Unsafe" />
+           </imports>
+        </module>
+	<module name="org.apache.lucene" slot="5.3.1"/>
```
these should use the version properties defined in the parent pom, it makes it easier to maintain that way

hchiorean commented on index-providers/modeshape-elasticsearch-index-provider/src/main/java/org/modeshape/jcr/index/elasticsearch/TransportAddress.java:
```diff
@@ -0,0 +1,35 @@
+ *
+ * @author kulikov
+ */
+public class TransportAddress {
```
what's the purpose of this class ? Hardcoding a port doesn't seem right...

okulikov replied:
This class is used for parsing addresses and ports of the ES cluster nodes (actually it is separated as dedicated class because it is easy to test its feature rather then implement/test as private method inside `EsIndexProvider` class). The default value is localhost:9300. No hardcoded  ports.

hchiorean replied:
u'r right, I didn't notice that constant is just the default value
still, I'm not 100% sure a separate class is required for testing essentially a `string.split` method

okulikov replied:
@hchiorean , Just to be sure. At this moment I can do a copy/paste of the method and discard the class.

hchiorean commented on index-providers/modeshape-elasticsearch-index-provider/src/test/java/org/modeshape/jcr/index/elasticsearch/EsIndexProviderTest.java:
```diff
@@ -0,0 +1,127 @@
+import org.modeshape.jcr.api.query.Query;
+
+/**
+ * Unit test for {@link LuceneIndexProvider}. Since this is a local provider in term of repository locality, we want to run
```
This JavaDoc should be fixed, it's not accurate.

hchiorean commented on integration/modeshape-jbossas-integration-tests/src/main/webapp/WEB-INF/jboss-deployment-structure.xml:
```diff
@@ -0,0 +1,10 @@
+    <deployment>
+	<dependencies>
+	    <module name="javax.jcr"/>
+	    <module name="org.elasticsearch" slot="2.1.0" services="import" />
```
this should use the Maven version property.

hchiorean commented on integration/modeshape-jbossas-integration-tests/src/main/webapp/WEB-INF/jboss-deployment-structure.xml:
```diff
@@ -0,0 +1,10 @@
+<jboss-deployment-structure>
```
why is this explicit dependency file required by the test ? 
The WF config specifies a module dependency in the actual index-provider element:
```
<index-provider name="elasticsearch" classname="org.modeshape.jcr.index.elasticsearch.EsIndexProvider" module="org.modeshape.index-provider.elasticsearch"/>

as such, the CP should be resolved at runtime correctly.

okulikov replied:
@hchiorean , This explicit dependency allows to start local ES node before running test

hchiorean commented:
@okulikov this looks good, the only thing that's left is using the `${}` version properties in the `module.xml` file
Also, can you please squash the 3 commits into 1

@rhauch are you ok with the PR ?

rhauch commented on deploy/jbossas/kit/jboss-wf9/org/apache/lucene/5.3.1/module.xml:
```diff
@@ -23,8 +23,16 @@
+	<resource-root path="lucene-suggest-${version.org.apache.lucene}.jar" />
+	<resource-root path="lucene-misc-${version.org.apache.lucene}.jar" />
+	<resource-root path="lucene-grouping-${version.org.apache.lucene}.jar" />
+	<resource-root path="lucene-spatial-${version.org.apache.lucene}.jar" />
```
Are all of these used in the index provider? Do they all come with the standard Lucene install, or were some added from other downloads? For example, I'm surprised if we use spatial or highlighter in our index provider.

rhauch replied:
Ah, never mind. I guess the Elastic Search client library probably requires them all, even if we don't directly use those aspects of Lucene.

rhauch commented on index-providers/modeshape-elasticsearch-index-provider/src/main/java/org/modeshape/jcr/index/elasticsearch/EsIndexColumn.java:
```diff
@@ -0,0 +1,254 @@
+            case URI:
+                return valueFactories.getStringFactory().create(value);
+            case DATE :
+                return ((DateTime) value).getMilliseconds();
```
There is no `default` case here. What about longs or binaries or other `PropertyType`s?

okulikov replied:
No conversation will happen. The last statement returns value as is

rhauch commented on index-providers/modeshape-elasticsearch-index-provider/src/main/java/org/modeshape/jcr/index/elasticsearch/EsIndexColumn.java:
```diff
@@ -0,0 +1,254 @@
+                return valueFactories.getSimpleReferenceFactory().create(value);
+            case WEAKREFERENCE :
+                return valueFactories.getWeakReferenceFactory().create(value);
+        }
```
Again, this switch is incomplete. IIUC, this should even compile with the checkstyle rules.

okulikov replied:
Same as above. If type is outside of the switch scope the same value will be returned.

rhauch replied:
Not having the switch include all possible values is very much preferred, and using the `default` case is a secondary option. However, doing neither is IMO bad form since it's implicitly providing a default case and not using Java's built-in `default` case. Yes, you'll need to throw an exception as the last line of the method to signal that getting there should never happen; even the ugliness of throwing an exception is better than an incomplete switch.

rhauch commented on index-providers/modeshape-elasticsearch-index-provider/src/main/java/org/modeshape/jcr/index/elasticsearch/Operations.java:
```diff
@@ -0,0 +1,351 @@
+            SetCriteria setCriteria = (SetCriteria)constraint;
+            
+            String field = (String) operand(setCriteria.getOperand()).apply(setCriteria.getOperand(), variables);
+            ArrayList list = new ArrayList();
```
We should never used untyped generics.

rhauch commented on index-providers/modeshape-elasticsearch-index-provider/src/main/java/org/modeshape/jcr/index/elasticsearch/Operations.java:
```diff
@@ -0,0 +1,351 @@
+            for (StaticOperand so : setCriteria.getValues()) {
+                Object vals = operand(so).apply(field, so, variables);
+                if (vals instanceof Object[]) {
+                    list.addAll(Arrays.asList((Object[]) vals));
```
The values will be of the type used within the query system. Are you sure these values don't need to be converted before they're passed into the query?

rhauch commented on index-providers/modeshape-elasticsearch-index-provider/src/main/java/org/modeshape/jcr/index/elasticsearch/Operations.java:
```diff
@@ -0,0 +1,351 @@
+        }
+    };
+
+    private final OperandBuilder LOWER_CASE_BUILDER = new OperandBuilder<LowerCase>() {
```
If these are immutable constants, they should be marked as static. If not, their name shouldn't be uppercased.

rhauch commented on index-providers/modeshape-elasticsearch-index-provider/src/main/java/org/modeshape/jcr/index/elasticsearch/SearchResults.java:
```diff
@@ -0,0 +1,92 @@
+                    .execute()
+                    .actionGet();
+        } catch (Exception e) {
+            e.printStackTrace();
```
Is there no better way to handle this exception? At least it should be logged and not printed.

rhauch commented on index-providers/modeshape-elasticsearch-index-provider/src/main/java/org/modeshape/jcr/index/elasticsearch/SearchResults.java:
```diff
@@ -0,0 +1,92 @@
+                    .actionGet();
+            return response.getHits().getTotalHits();
+        } catch (Exception e) {
+            return 0;
```
So if there is a network error, this method will return `0`? Seems strange.

rhauch commented on index-providers/modeshape-elasticsearch-index-provider/src/test/java/org/modeshape/jcr/index/elasticsearch/EsIndexProviderTest.java:
```diff
@@ -0,0 +1,126 @@
+    @Test
+    @Override
+    public void shouldUseIndexesAfterRestarting() throws Exception {
+    }
```
Can this test not be implemented or at least specialized so that we know that the index provider after restart will connect properly to a still-running Elasticsearch node?

okulikov replied:
Local ES node used for testing switches to "YELLOW" state during repository restart (I guess this is ES bug )so we don't have still-running ES node during this test.

rhauch replied:
Then perhaps add a comment stating such?

okulikov replied:
makes sense

rhauch commented:
@okulikov, @hchiorean: see my questions above.

okulikov commented:
@rhauch , @hchiorean Fixed commented lines

hchiorean commented:
@okulikov: I'm running the integrated PR code on my machine and there are some things which have to be fixed:

1. https://gist.github.com/hchiorean/c501770624ee74775438 - a lot of javac warnings due unchecked/raw calls. If any/some of those calls are due to the ES API, then they should be avoided with `SupressWarnings`. Otherwise generic types should be used.

2. the time required to run the `EsIndexProviderTest` on my machine is extremely high: `7 minutes`. This is something which has to be fixed, because IMO it's simply not feasible to add a module in the MS general build which increases the build duration by this much.
For example, in the `EsIndexProviderTest` class, why is an `org.elasticsearch.node.Node` started and stopped with each test ? This doesn't seem right: the ES cluster should be started once before all the tests and stopped after all the tests (i.e. `BeforeClass` and `AfterClass`). This would also eliminate the need to `Thread.sleep(500)` between each test, just to wait for a shutdown

3. the integration tests are failing: https://gist.github.com/hchiorean/33cdc73c23ebbe5265dc meaning that the WF integration is not correct

4. ES client - this is something which hasn't been discussed so far, but it should be: 
the only thing ModeShape should require for this integration is the ES (or another) API that *communicates with the cluster* and nothing else. In other words, *no dependency from the following graph should be present in the ModeShape ES distributions*: 
```
 +- org.apache.lucene:lucene-core:jar:5.3.1:compile
 +- org.apache.lucene:lucene-backward-codecs:jar:5.3.1:compile
 +- org.apache.lucene:lucene-analyzers-common:jar:5.3.1:compile
 +- org.apache.lucene:lucene-queries:jar:5.3.1:compile
 +- org.apache.lucene:lucene-memory:jar:5.3.1:compile
 +- org.apache.lucene:lucene-highlighter:jar:5.3.1:compile
 +- org.apache.lucene:lucene-queryparser:jar:5.3.1:compile
 |  \- org.apache.lucene:lucene-sandbox:jar:5.3.1:compile
 +- org.apache.lucene:lucene-suggest:jar:5.3.1:compile
 |  \- org.apache.lucene:lucene-misc:jar:5.3.1:compile
 +- org.apache.lucene:lucene-join:jar:5.3.1:compile
 |  \- org.apache.lucene:lucene-grouping:jar:5.3.1:compile
 +- org.apache.lucene:lucene-spatial:jar:5.3.1:compile
 |  +- org.apache.lucene:lucene-spatial3d:jar:5.3.1:compile
 |  \- com.spatial4j:spatial4j:jar:0.4.1:compile
 +- com.google.guava:guava:jar:18.0:compile
 +- com.carrotsearch:hppc:jar:0.7.1:compile
 +- joda-time:joda-time:jar:2.8.2:compile
 +- org.joda:joda-convert:jar:1.2:compile
 +- com.fasterxml.jackson.core:jackson-core:jar:2.6.2:compile
 +- com.fasterxml.jackson.dataformat:jackson-dataformat-smile:jar:2.6.2:compile
 +- com.fasterxml.jackson.dataformat:jackson-dataformat-yaml:jar:2.6.2:compile
 |  \- org.yaml:snakeyaml:jar:1.15:compile
 +- com.fasterxml.jackson.dataformat:jackson-dataformat-cbor:jar:2.6.2:compile
 +- io.netty:netty:jar:3.10.5.Final:compile
 +- com.ning:compress-lzf:jar:1.0.2:compile
 +- com.tdunning:t-digest:jar:3.0:compile
 +- org.hdrhistogram:HdrHistogram:jar:2.1.6:compile
 +- commons-cli:commons-cli:jar:1.3.1:compile
 \- com.twitter:jsr166e:jar:1.1.0:compile
```
@rhauch touches on this topic in his previous comment

The client should only send/receive requests and therefore not require any of the above (of course we might need the full lot for testing, but otherwise none should be required). Now, I'm also aware (see https://www.elastic.co/blog/found-java-clients-for-elasticsearch) that ES doesn't provide a dedicated client jar OOTB (which truly sucks IMO). If that means we need to bring in all of the above dependencies, then we need another client. The previous link lists some possible Java alternatives.

One of the main reasons why having this lot of dependencies is not an option, is that this means the ES index provider *will never be productized* since many of those dependencies will most likely never be compatible with the ones from EAP 7.

In conclusion, the MS ES index provider should use a client API which is as lightweight as possible and which doesn't bring in a lot of transitive dependencies.

okulikov commented:
@hchiorean , I also extremely dislike such long list of dependencies not transparent for me as well as dependencies from the `Lucene` and 'Elasticsearch' jars. And if we will agree I would prefer to implement a simple and lightweight client for our purposes

hchiorean commented:
@okulikov +1 

Ideally we wouldn't implement everything from scratch. Perhaps we can use some code from/fork https://github.com/searchbox-io/Jest. Looking at it the only dependency we don't need from their codebase is `guava` and perhaps `gson`. Apart from those 2 it's lightweight enough and ALS2 licensed.

@rhauch any thoughts on this topic ? (point 4 from my previous comment)

rhauch commented:
@hchiorean, @okulikov: is it possible to use the official client and prune nearly all of the dependencies? If we're not using much of that code, then might this work?

okulikov commented:
@rhauch, @hchiorean  it is not possible. we already are using minimal code and it still depend from all that jars. Also I am thinking that we should not shy to have our own lightweight client. What I understood is that the most "complicated" part is reading and parsing response but in our case it is simple because we need to extract only identifier of the node. Making request suppose only string formating.

rhauch commented:
@okulikov, I'm fine with trying out our own client, since the requirements are quite narrowly focused and constrained.

okulikov commented:
@hchiorean , @rhauch, Latest commit contains lightweight ES client which does not depend from any external jars now and works fast. Dependency from `elasticsearch` jar remains only in test scope and it is required to start ES cluster only for unit-tests. 
However, for integration test the problem still remain. What I do is
- create arquillian test case;
- create war deployment using ShrinkWrap where WEB-INF/lib folder contains `elasticsearch-2.1.0` jar with transitive dependencies  `lucene-*-5.3.1.jar`;
- `@before` test-case ES node started, `@after` test-case stopped.

`War` itself looks good. All necessary classes are really in the `WEB-INF/lib` folder.  However ES node doesn't start. Exception indicates that there is a version mess. Reason is that WF provides its own `lucene-4.10` unit with older classes and that classes are on the same class path with web application. I tried all tricks with `manifest.mf`, `jboss-deployment-descriptor`, repackaging... The only working way is to define custom module, put all jars to that module and explicit specify dependency. But this is weird:
- distribution contains custom module which is used only for integration test;
- this custom module contains the same long list of jars.

Another options:
- using [ES testing framework](https://www.elastic.co/guide/en/elasticsearch/reference/2.x/testing-framework.html); It is not available in the current release (only on snapshots);
- using `Docker`; This way looks overcomplicated for running one simple IT test case.

Taking into account that test-case itself enough simple (index to nodes and search with simple query) my suggestion is to implement also our own simple backing node. All the ES specific functionality covered by functional unit-test and the purpose of IT test is that provider is plugged and communication exists. What do you think guys?

hchiorean commented:
@okulikov: did you try explicitly excluding the older Lucene version ? The version mess is caused by ISPN 7.x bring in its own version of Lucene & Co. Fortunately, these are installed in separate AS modules, so one option is to tweak the `jboss-deployment-descriptor` of the ShrinkWrap WAR and explicitly remove those modules.

see https://developer.jboss.org/message/850892#850892 for an example of dependency descriptor which excludes certain modules

okulikov commented:
@hchiorean , let me try

okulikov commented:
@hchiorean, Declarative exclusions did not help. I also tried to pick up dependencies from `ShrikWrap` and load required classes with custom class loader and also without any luck. 
In latest suggested  another way - prepare and deploy elasticsearch WF module using maven especially for integration tests only. Outside of the integration test this module does not exists.

hchiorean commented on index-providers/modeshape-elasticsearch-index-provider/src/test/java/org/modeshape/jcr/index/elasticsearch/EsIndexProviderTest.java:
```diff
@@ -0,0 +1,132 @@
+    public void afterEach() throws Exception {
+        try {
+            stopRepository();
+        } catch (Exception e) {
```
should this exception be left to bubble up and fail the test ? If `afterEach` fails it means the repository was not shutdown properly

hchiorean commented on index-providers/modeshape-elasticsearch-index-provider/src/test/java/org/modeshape/jcr/index/elasticsearch/EsIndexTest.java:
```diff
@@ -0,0 +1,369 @@
+        try {
+            Thread.currentThread().sleep(1000);
+        } catch (Exception e) {
+        }
```
the only possible checked exception here is `InterruptedException` which if thrown, should be handled accordingly (i.e. interrupted status cleared and perhaps rethrown)

hchiorean commented on index-providers/modeshape-elasticsearch-index-provider/src/test/java/org/modeshape/jcr/index/elasticsearch/EsIndexTest.java:
```diff
@@ -0,0 +1,369 @@
+    }
+
+    @Before
+    public void setUp() {
```
why have  this method at all if it's empty ?

hchiorean commented on index-providers/modeshape-elasticsearch-index-provider/src/test/java/org/modeshape/jcr/index/elasticsearch/EsIndexTest.java:
```diff
@@ -0,0 +1,369 @@
+    }
+
+    @After
+    public void tearDown() {
```
why have  this method at all if it's empty ?

hchiorean commented on index-providers/modeshape-elasticsearch-index-provider/src/test/java/org/modeshape/jcr/index/elasticsearch/EsIndexTest.java:
```diff
@@ -0,0 +1,369 @@
+        try {
+            Thread.currentThread().sleep(1000);
+        } catch (Exception e) {
+        }
```
again, interruption should not be ignored

hchiorean commented on integration/modeshape-jbossas-integration-tests/pom.xml:
```diff
@@ -231,6 +241,20 @@
+                        <goals>
+                            <goal>copy-dependencies</goal>
+                        </goals>                        
+                        <configuration>
```
shouldn't this be configured to only copy the `org.elasticsearch` artifact-related dependencies ? In other words, don't copy all the module dependencies. Also, the generic `libs` name can then be changed to smth like `elasticsearchLibs` or something similar

okulikov replied:
@hchiorean , in case of artifact-related configuration maven ignores transitive dependencies. Later, for WF module I need to specify jar names so for this particular case more useful to follow rule based on jar names. 

I did change dependency output directory name from generic `lib` to more specific `eslib`

hchiorean commented on integration/modeshape-jbossas-integration-tests/pom.xml:
```diff
@@ -259,6 +283,45 @@
+                                    <include>module.xml</include>
+                                </resource>
+                                <resource>
+                                    <directory>${project.build.directory}/libs</directory>
```
if the above `copy-dependencies` goal is configured to only include `elasticsearch` related artifacts, this won't need this entire list of specific artifacts to include. Instead, in can include everything from the `elasticsearchLibs` dir.

hchiorean commented:
@okulikov: this looks ok to me with some minor comments (see above)

rhauch commented:
@okulikov: I agree with @hchiorean: those smaller issues should be addressed, and then this looks good.

okulikov commented:
@hchiorean , @rhauch, New commit with fixes for the commented issues.

hchiorean commented:
@okulikov: the ES integration tests are failing on my machine when trying to build. See https://gist.github.com/hchiorean/a9ff8edc7194e1c3a821 for more details

okulikov commented:
@hchiorean , can you please more lines from above. It is not visible from the gist status of ES cluster.
Exception shows that ES cluster is not active.

hchiorean commented:
@okulikov a couple of things:

1.  the `Connection refused: connect` exception is raised *not* when running the `QueryIntegrationTest` but `DatasourceIntegrationTest`. See [this gist](https://gist.github.com/hchiorean/a9ff8edc7194e1c3a821) which is the full output from the console. The reason is simple: the ES index is defined as part of the `artifacts` repository which is used from different test cases. However, only the  `QueryIntegrationTest` contains the necessary boiler plating for starting/stopping the ES cluster. As such, the ES index integration test _should use a separate, dedicated repository_, not one which is shared by other test cases

2.  the test failure from `QueryIntegrationTest` may or may not be related to the above. However if you look at [this gist](https://gist.github.com/hchiorean/d796cf1eed1abe4c1b00) (which is again the full console output) you can see that the ES cluster starts up correctly but the test fails with a regular failure.

maybe correcting (1) will fix (2), but the fix need to be tested by running all the tests from the `modeshape-integration-test` module.

okulikov commented:
@hchiorean , Agree with (1). Let me fix it fist.

okulikov commented:
@hchiorean , updated commit. Moved Elasticsearch index test to a dedicated test case and repository. Also I added 3 seconds delay because the only visible reason for (2) failure is thread racing. On my side it runs without failure constantly.

hchiorean commented:
@okulikov: the test also passes/passed on my machine when ran in isolation. However, it failed constantly when running `mvn clean install -Pintegration` on the project root.

hchiorean commented on integration/modeshape-jbossas-integration-tests/src/test/java/org/modeshape/test/integration/QueryIntegrationTest.java:
```diff
@@ -54,14 +54,15 @@
-                         .addAsResource(new File("src/test/resources/text-extractor/text-file.txt"))
-                         .setManifest(new File("src/main/webapp/META-INF/MANIFEST.MF"));
+                .addAsWebInfResource(EmptyAsset.INSTANCE, ArchivePaths.create("beans.xml"))
+                .addAsWebInfResource(new File("src/main/webapp/WEB-INF/jboss-deployment-structure.xml"), ArchivePaths.create("jboss-deployment-structure.xml"))
```
since the ES test have been moved to another class, this isn't required here, right ?

hchiorean commented on integration/modeshape-jbossas-integration-tests/src/test/java/org/modeshape/test/integration/EsIndexIntegrationTest.java:
```diff
@@ -0,0 +1,147 @@
+
+        //configure Elasticsearch node
+        esNode = NodeBuilder.nodeBuilder().settings(localSettings).local(false).build().start();
+        Thread.currentThread().sleep(3000);
```
I think 1000 (1sec) is enough here since this is the same value that the standalone tests use (from the es-index-provoider module)

hchiorean commented:
@okulikov: with the latest changes the build runs fine; I just had a couple of minor test comments

okulikov commented:
@hchiorean , @rhauch , Updated with accordance to latest comments.

hchiorean commented:
@okulikov the build still passes, so it looks good
@rhauch are you ok with merging this ?

rhauch commented on index-providers/modeshape-elasticsearch-index-provider/src/main/java/org/modeshape/jcr/index/elasticsearch/EsIndex.java:
```diff
@@ -0,0 +1,308 @@
+            EsDocument doc = findOrCreateDoc(nodeKey);
+            putValues(doc, column, values);
+            client.storeDocument(name(), workspace, nodeKey, doc);
+            client.refresh(name());
```
@okulikov, can you explain why `client.refresh(...)` is called after every add and remove rather than during the commit of changes to the index? Perhaps I don't understand the API well enough, but ES refresh seems to cause a refresh of the index so that any changes since the last refresh are immediately visible.

rhauch replied:
BTW, I know that `ESClient` is ModeShape-specific, but it seems to be a convenient wrapper around the normal ES API.

rhauch commented on index-providers/modeshape-elasticsearch-index-provider/src/main/java/org/modeshape/jcr/index/elasticsearch/EsIndex.java:
```diff
@@ -0,0 +1,308 @@
+    @Override
+    public void commit() {
+        try {
+            client.flush(name());
```
Again, perhaps I don't understand the Java API, but IIUC `client.flush(...)` is explicitly telling ES to flush all memory associated with an index to storage, and that ES tries to do this automatically. If we have to do this, then `commit()` is definitely the place to do it. But is it really required? Do we want to force a flush to get the durability of the write to disk?

rhauch commented on index-providers/modeshape-elasticsearch-index-provider/src/main/java/org/modeshape/jcr/index/elasticsearch/client/EsDocument.java:
```diff
@@ -0,0 +1,133 @@
+ * @author kulikov
+ */
+public class EsDocument {
+    private EditableDocument document = DocumentFactory.newDocument();
```
The initialization of `document` should be moved into the default constructor so the other constructor doesn't do it twice. Not sure how much of an impact it will have, but its simple enough to correct.

rhauch commented on index-providers/modeshape-elasticsearch-index-provider/src/main/java/org/modeshape/jcr/index/elasticsearch/client/EsDocument.java:
```diff
@@ -0,0 +1,133 @@
+           }
+           return res;
+        }
+        return document.get(name);
```
Shouldn't this just return `obj`?

rhauch commented on index-providers/modeshape-elasticsearch-index-provider/src/main/java/org/modeshape/jcr/index/elasticsearch/client/EsDocument.java:
```diff
@@ -0,0 +1,133 @@
+ * 
+ * @author kulikov
+ */
+public class EsDocument {
```
I'm curious why this class even exists. It seems like it is less efficient in most read-only cases, since this class wraps an _editable_ document, which may not be needed when the EsDocument is just read.

rhauch commented on index-providers/modeshape-elasticsearch-index-provider/src/main/java/org/modeshape/jcr/index/elasticsearch/client/EsDocument.java:
```diff
@@ -0,0 +1,133 @@
+    public Object get(String name) {
+        Object obj = document.get(name);
+        if (obj instanceof BasicArray) {
+           Object[] res = new Object[((BasicArray)obj).size()];
```
Perhaps `Array` (which `BasicArray` implements) also implements `List<Object>`, so it should be possible to just call `array.toArray(new Object[array.size()])`. Perhaps `Array` should provide a more convenient method for doing exactly this.

Although, I still think that if the ES index provider code directly used `Document` and `Array`, this would not really be needed.

okulikov commented:
@rhauch , PR updated:
- Refresh process moved to index `commit`method; You are right, no need to call each time for add/remove;
- `EsDocument` replaced with two classes dedicated for read and write; 
- `arrays` conversation with more suitable method.

`EsIndex` does all required types conversations using information given by `IndexColumnDefinition` and value factories provided by MS (value factories receive `Object` argument). In case of multi-value property such conversion is applied for each element of an `Object[]` array. Thus there is no need to get/set typed value but It seems to me that it is convenient to introduce wrapper(s) around `Document` and hide a differences between array types (ie `Object[]` and `BasicArray`)

rhauch commented:
Looks good to me.

hchiorean commented:
@okulikov: I merged this in `master`, but it does not cherry-pick cleanly in `4.x`.
Can you pls cherry-pick & fix conflicts and then create a separate PR for `4.x` ? Thanks.
